Ruby backend

danielsdeleo opened the pull request:
In this pull request, I've added a ruby based solver option using [molinillo](https://github.com/CocoaPods/Molinillo) which Bundler now uses. In the long term, my goal is that when I use solve with ChefDK for policyfiles, we can just use the ruby solver, which will save me the pain of having to maintain the gecode build on a lot of different platforms.

For berks, the issue is much more complicated, of course. For users who use Chef Solo (or local mode) or environment cookbooks or some other homegrown lockdown mechanism, the exact ranking of different solutions isn't too big a deal because they're guaranteed to deploy what they test. Unfortunately, many users don't use any lockdown mechanism, and will depend on berks generating the same solution as the chef server would.

On the other hand, there are users on platforms for which Chef doesn't ship a prebuilt ChefDK or gem install anyway for whatever reason. These users would benefit if we gave them the option to just use the pure-ruby solver.

FWIW, the ruby solver seems to actually be a little faster than the gecode one on the benchmark we have:

```
$ bundle exec ruby spec/acceptance/benchmark.rb
                   user     system      total        real
Create graph   0.200000   0.000000   0.200000 (  0.196406)
Solve Gecode   0.120000   0.010000   0.130000 (  0.130030)
Solve Ruby     0.090000   0.000000   0.090000 (  0.088621)
```

But of course dependency solving is a complicated problem so who knows how the backends might compare on different data.

Anyway, thoughts?

reset commented:
@danielsdeleo a ruby based solver implementing Molinillo is greatly preferred to the gecode solution we currently have. Using anything other than what the Chef Server is using wouldn't be a good idea, though, since the Chef Server is deterministically solving some complicated graphs incorrectly.

I don't think this would be much of a problem moving forward since the burden of resolution will be removed from the Chef Server, right?

danielsdeleo commented:
@reset Yeah, if people use policyfiles or any of the existing lockdown mechanisms based on environments (or the really weird stuff with auto generated cookbook metadata) they're avoiding use of the server side solver, so how we weight the possible solutions doesn't matter.

From a Chef Server maintenance policy perspective, it's gonna be quite a long time until we can drop the server-side dependency solver. Policyfiles are just landing on hosted and non-feature-flagged in Chef Server some time in the next week (if everything goes according to plan), and lots of folks are still on various 11.x versions of Client/Server. Eventually if we can get everyone on to some workflow that doesn't hit the solver (which we really want to do), then we'll be able to remove it.

For right now I'm wondering if it would be optimal to have the backend be selectable and make berks opt for the gecode solver, only if available, by default. So a gem install of berks would use the ruby based solver until you install dep_selector, but in ChefDK we'll continue to build and ship dep_selector.

reset commented:
@danielsdeleo ok that sounds great to me. Let's go ahead and have the Ruby based solver be the default if gecode is not present. Here's a few additional things I think is worth mentioning:

* We should allow the solver to be explicitly configurable by extending the Berksfile DSL
* The DSL addition will allow you to configure a solver as `:preferred` or `:required`
* If the solver isn't available on the client's machine and the solver setting is set to `:preferred` we should display a warning
* If the solver isn't available on the client's machine and the solver setting is set to `:required` we should cancel execution and exit with a non-zero exit code

danielsdeleo commented:
Sounds like an awesome plan. Setting the desired solver in the berksfile sounds like a really good idea.

reset commented:
@danielsdeleo :+1: great, let me know if you need anything from me!

danielsdeleo commented:
@reset I made some more progress on this. The acceptance tests for the ruby solver will have some failures, but I wanted to note a few things:

* I changed the expected text for solve failures to just expect what molinillo produces. It uses Solve's error classes though.
* The solutions it gets are indeed different from what gecode does, but I think generally they're preferable; for example, this acceptance test gets the same result as the old ruby solver did: https://github.com/berkshelf/solve/commit/14b17722b552b0d3e4fa061ece5d9c7baddf7009#diff-8b23e53518f1346aa2ea3ec477634046R243
* Molinillo doesn't accept circular dependencies at all. You should double-check that the way I changed this test is ok: https://github.com/berkshelf/solve/commit/14b17722b552b0d3e4fa061ece5d9c7baddf7009#diff-8b23e53518f1346aa2ea3ec477634046R191 Also, circular dep errors are the cause of the remaining failures in the acceptance tests.
* I made the ruby solver the default, which probably means we should do a major version bump when this is merged.

Aside from the issues with the acceptance tests, I think the code is generally ready for review.

danielsdeleo commented:
One more thing, the way I wrote passing options into the solver, where they go both to `new` and `#resolve` is not great. We definitely need to pass options in to the ruby solver, it uses them in error messages. For example, the default error would be:

```
`yum (= 2.4.5)` required by `user-specified dependency`
```

But you can make it instead say:

```
`yum (= 2.4.5)` required by `Berksfile`
```

danielsdeleo commented:
Also, doing some more testing with the benchmarks. For the `LargeGraphNoSolution` case, the ruby solver takes ~ 1 second and the error looks like:

```
Unable to satisfy the following requirements:

- `yum (= 2.4.5)` required by `user-specified dependency`
- `yum (~> 3.1)` required by `virtualbox-cookbook-1.0.3`
- `yum (>= 0.0.0)` required by `logstash-0.7.6`
- `yum (~> 3.0)` required by `jenkins-2.0.3`
- `yum (~> 3.0)` required by `memcached-1.7.2`
- `yum (~> 3.0)` required by `yum-epel-0.3.4`
- `yum (< 3.0.0)` required by `git-2.8.5`
- `yum (~> 3.0)` required by `git-2.8.5`
```

Whereas the gecode solver takes 4 seconds and produces this error message:

```
Unable to satisfy constraints on package yum due to solution constraint (jenkins = 2.0.3). Solution constraints that may result in a constraint on yum: [(sensu = 0.1.0) -> (rabbitmq = 3.1.0) -> (erlang = 1.5.2) -> (yum ~> 3.0)], [(sensu = 0.1.0) -> (rabbitmq = 3.1.0) -> (erlang = 1.5.2) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(sensu = 0.1.0) -> (rabbitmq = 3.1.0) -> (erlang = 1.5.2) -> (yum-erlang_solutions = 0.2.0) -> (yum ~> 3.0)], [(sensu = 0.1.0) -> (rabbitmq = 3.1.0) -> (erlang = 1.3.6) -> (yum < 3.0.0)], [(sensu = 0.1.0) -> (rabbitmq = 3.1.0) -> (erlang = {1.3.2,1.2.0,1.1.2}) -> (yum >= 0.5.0)], [(sensu = 0.1.0) -> (rabbitmq = 1.8.0) -> (yum >= 0.5.0)], [(sensu = 0.1.0) -> (rabbitmq = 1.8.0) -> (erlang = 1.5.2) -> (yum ~> 3.0)], [(sensu = 0.1.0) -> (rabbitmq = 1.8.0) -> (erlang = 1.5.2) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(sensu = 0.1.0) -> (rabbitmq = 1.8.0) -> (erlang = 1.5.2) -> (yum-erlang_solutions = 0.2.0) -> (yum ~> 3.0)], [(sensu = 0.1.0) -> (rabbitmq = 1.8.0) -> (erlang = 1.3.6) -> (yum < 3.0.0)], [(sensu = 0.1.0) -> (rabbitmq = 1.8.0) -> (erlang = {1.3.2,1.2.0,1.1.2}) -> (yum >= 0.5.0)], [(sensu = 0.1.0) -> (rabbitmq = 1.4.0) -> (yum >= 0.5.0)], [(sensu = 0.1.0) -> (rabbitmq = 1.4.0) -> (erlang = 1.5.2) -> (yum ~> 3.0)], [(sensu = 0.1.0) -> (rabbitmq = 1.4.0) -> (erlang = 1.5.2) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(sensu = 0.1.0) -> (rabbitmq = 1.4.0) -> (erlang = 1.5.2) -> (yum-erlang_solutions = 0.2.0) -> (yum ~> 3.0)], [(sensu = 0.1.0) -> (rabbitmq = 1.4.0) -> (erlang = 1.3.6) -> (yum < 3.0.0)], [(sensu = 0.1.0) -> (rabbitmq = 1.4.0) -> (erlang = {1.3.2,1.2.0,1.1.2}) -> (yum >= 0.5.0)], [(sensu = 0.1.0) -> (redis = 3.0.4) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(sensu = 0.1.0) -> (redis = 3.0.4) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(sensu = 0.1.0) -> (redis = 3.0.4) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(sensu = 0.1.0) -> (redis = 3.0.4) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(sensu = 0.1.0) -> (redis = 3.0.0) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(sensu = 0.1.0) -> (redis = 3.0.0) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(sensu = 0.1.0) -> (redis = 3.0.0) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(sensu = 0.1.0) -> (redis = 3.0.0) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(sensu = 0.1.0) -> (nginx = 2.5.1) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(sensu = 0.1.0) -> (nginx = 2.5.1) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(sensu = 0.1.0) -> (nginx = 2.5.1) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(sensu = 0.1.0) -> (nginx = 2.5.1) -> (runit = 1.4.0) -> (yum >= 0.0.0)], [(sensu = 0.1.0) -> (nginx = 2.5.1) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(sensu = 0.1.0) -> (nginx = {1.2.0,1.6.0}) -> (yum >= 0.0.0)], [(sensu = 0.1.0) -> (nginx = 1.6.0) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(sensu = 0.1.0) -> (nginx = 1.6.0) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(sensu = 0.1.0) -> (nginx = 1.6.0) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(sensu = 0.1.0) -> (nginx = 1.6.0) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(sensu = 0.1.0) -> (nginx = 2.4.2) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(sensu = 0.1.0) -> (nginx = 2.4.2) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(sensu = 0.1.0) -> (nginx = 2.4.2) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(sensu = 0.1.0) -> (nginx = 2.4.2) -> (runit = 1.4.0) -> (yum >= 0.0.0)], [(sensu = 0.1.0) -> (nginx = 2.4.2) -> (yum ~> 3.0)], [(sensu = 0.1.0) -> (nginx = 2.4.2) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(sensu = 0.1.0) -> (nginx = 2.0.8) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(sensu = 0.1.0) -> (nginx = 2.0.8) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(sensu = 0.1.0) -> (nginx = 2.0.8) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(sensu = 0.1.0) -> (nginx = 2.0.8) -> (runit = 1.4.0) -> (yum >= 0.0.0)], [(sensu = 0.1.0) -> (nginx = 2.0.8) -> (yum < 3.0.0)], [(sensu = 0.1.0) -> (nginx = 2.0.4) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(sensu = 0.1.0) -> (nginx = 2.0.4) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(sensu = 0.1.0) -> (nginx = 2.0.4) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(sensu = 0.1.0) -> (nginx = 2.0.4) -> (runit = 1.4.0) -> (yum >= 0.0.0)], [(sensu = 0.1.0) -> (nginx = 2.0.4) -> (yum ~> 2.3)], [(sensu = 0.1.0) -> (nginx = 1.4.0) -> (yum >= 0.0.0)], [(sensu = 0.1.0) -> (nginx = 1.0.2) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(sensu = 0.1.0) -> (nginx = 1.0.2) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(sensu = 0.1.0) -> (nginx = 1.0.2) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(sensu = 0.1.0) -> (nginx = 1.0.2) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(sensu = 0.1.0) -> (nginx = 1.0.2) -> (yum >= 0.0.0)], [(sensu = 0.1.0) -> (nginx = 1.0.0) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(sensu = 0.1.0) -> (nginx = 1.0.0) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(sensu = 0.1.0) -> (nginx = 1.0.0) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(sensu = 0.1.0) -> (nginx = 1.0.0) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(sensu = 0.1.0) -> (nginx = 1.0.0) -> (yum >= 0.0.0)], [(sensu = 0.1.0) -> (nginx = 0.101.2) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(sensu = 0.1.0) -> (nginx = 0.101.2) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(sensu = 0.1.0) -> (nginx = 0.101.2) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(sensu = 0.1.0) -> (nginx = 0.101.2) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(sensu = 0.1.0) -> (nginx = 0.100.2) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(sensu = 0.1.0) -> (nginx = 0.100.2) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(sensu = 0.1.0) -> (nginx = 0.100.2) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(sensu = 0.1.0) -> (nginx = 0.100.2) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(sensu = 0.1.0) -> (nginx = 0.99.2) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(sensu = 0.1.0) -> (nginx = 0.99.2) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(sensu = 0.1.0) -> (nginx = 0.99.2) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(sensu = 0.1.0) -> (nginx = 0.99.2) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(sensu = 0.1.0) -> (nginx = 0.14.3) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(sensu = 0.1.0) -> (nginx = 0.14.3) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(sensu = 0.1.0) -> (nginx = 0.14.3) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(sensu = 0.1.0) -> (nginx = 0.14.3) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.8) -> (python = 1.4.7) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.8) -> (python = 1.4.4) -> (yum >= 0.0.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.8) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.8) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.8) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.8) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.8) -> (memcached = 1.7.2) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.8) -> (memcached = 1.7.2) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.8) -> (memcached = 1.7.2) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.8) -> (memcached = 1.7.2) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.8) -> (memcached = 1.7.2) -> (yum ~> 3.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.8) -> (memcached = 1.7.2) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.8) -> (memcached = 1.6.6) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.8) -> (memcached = 1.6.6) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.8) -> (memcached = 1.6.6) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.8) -> (memcached = 1.6.6) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.8) -> (memcached = 1.6.6) -> (yum < 3.0.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.8) -> (memcached = 1.6.2) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.8) -> (memcached = 1.6.2) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.8) -> (memcached = 1.6.2) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.8) -> (memcached = 1.6.2) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.8) -> (memcached = {1.6.2,1.2.0}) -> (yum >= 0.0.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.8) -> (memcached = 1.1.2) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.8) -> (memcached = 1.1.2) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.8) -> (memcached = 1.1.2) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.8) -> (memcached = 1.1.2) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.2) -> (python = 1.4.7) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.2) -> (python = 1.4.4) -> (yum >= 0.0.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.2) -> (memcached = 1.7.2) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.2) -> (memcached = 1.7.2) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.2) -> (memcached = 1.7.2) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.2) -> (memcached = 1.7.2) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.2) -> (memcached = 1.7.2) -> (yum ~> 3.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.2) -> (memcached = 1.7.2) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.2) -> (memcached = 1.6.6) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.2) -> (memcached = 1.6.6) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.2) -> (memcached = 1.6.6) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.2) -> (memcached = 1.6.6) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.2) -> (memcached = 1.6.6) -> (yum < 3.0.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.2) -> (memcached = 1.6.2) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.2) -> (memcached = 1.6.2) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.2) -> (memcached = 1.6.2) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.2) -> (memcached = 1.6.2) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.2) -> (memcached = {1.6.2,1.2.0}) -> (yum >= 0.0.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.2) -> (memcached = 1.1.2) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.2) -> (memcached = 1.1.2) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.2) -> (memcached = 1.1.2) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.2) -> (memcached = 1.1.2) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.0) -> (python = 1.4.7) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.0) -> (python = 1.4.4) -> (yum >= 0.0.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.0) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.0) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.0) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.0) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.0) -> (memcached = 1.7.2) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.0) -> (memcached = 1.7.2) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.0) -> (memcached = 1.7.2) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.0) -> (memcached = 1.7.2) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.0) -> (memcached = 1.7.2) -> (yum ~> 3.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.0) -> (memcached = 1.7.2) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.0) -> (memcached = 1.6.6) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.0) -> (memcached = 1.6.6) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.0) -> (memcached = 1.6.6) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.0) -> (memcached = 1.6.6) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.0) -> (memcached = 1.6.6) -> (yum < 3.0.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.0) -> (memcached = 1.6.2) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.0) -> (memcached = 1.6.2) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.0) -> (memcached = 1.6.2) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.0) -> (memcached = 1.6.2) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.0) -> (memcached = {1.6.2,1.2.0}) -> (yum >= 0.0.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.0) -> (memcached = 1.1.2) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.0) -> (memcached = 1.1.2) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.0) -> (memcached = 1.1.2) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(ganglia = 0.2.1) -> (graphite = 0.4.0) -> (memcached = 1.1.2) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(ganglia = 0.2.1) -> (graphite = 0.3.0) -> (python = 1.4.7) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(ganglia = 0.2.1) -> (graphite = 0.3.0) -> (python = 1.4.4) -> (yum >= 0.0.0)], [(ganglia = 0.2.1) -> (graphite = 0.3.0) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(ganglia = 0.2.1) -> (graphite = 0.3.0) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(ganglia = 0.2.1) -> (graphite = 0.3.0) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(ganglia = 0.2.1) -> (graphite = 0.3.0) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(ganglia = 0.2.1) -> (graphite = 0.1.2) -> (python = 1.4.7) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(ganglia = 0.2.1) -> (graphite = 0.1.2) -> (python = 1.4.4) -> (yum >= 0.0.0)], [(ganglia = 0.2.1) -> (graphite = 0.0.5) -> (python = 1.4.7) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(ganglia = 0.2.1) -> (graphite = 0.0.5) -> (python = 1.4.4) -> (yum >= 0.0.0)], [(ganglia = 0.2.1) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(ganglia = 0.2.1) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(ganglia = 0.2.1) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(ganglia = 0.2.1) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(git = 2.8.5) -> (yum < 3.0.0)], [(git = 2.8.5) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(git = 2.8.5) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(git = 2.8.5) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(git = 2.8.5) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(graphite = 0.4.8) -> (python = 1.4.7) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(graphite = 0.4.8) -> (python = 1.4.4) -> (yum >= 0.0.0)], [(graphite = 0.4.8) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(graphite = 0.4.8) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(graphite = 0.4.8) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(graphite = 0.4.8) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(graphite = 0.4.8) -> (memcached = 1.7.2) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(graphite = 0.4.8) -> (memcached = 1.7.2) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(graphite = 0.4.8) -> (memcached = 1.7.2) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(graphite = 0.4.8) -> (memcached = 1.7.2) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(graphite = 0.4.8) -> (memcached = 1.7.2) -> (yum ~> 3.0)], [(graphite = 0.4.8) -> (memcached = 1.7.2) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(graphite = 0.4.8) -> (memcached = 1.6.6) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(graphite = 0.4.8) -> (memcached = 1.6.6) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(graphite = 0.4.8) -> (memcached = 1.6.6) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(graphite = 0.4.8) -> (memcached = 1.6.6) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(graphite = 0.4.8) -> (memcached = 1.6.6) -> (yum < 3.0.0)], [(graphite = 0.4.8) -> (memcached = 1.6.2) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(graphite = 0.4.8) -> (memcached = 1.6.2) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(graphite = 0.4.8) -> (memcached = 1.6.2) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(graphite = 0.4.8) -> (memcached = 1.6.2) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(graphite = 0.4.8) -> (memcached = {1.6.2,1.2.0}) -> (yum >= 0.0.0)], [(graphite = 0.4.8) -> (memcached = 1.1.2) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(graphite = 0.4.8) -> (memcached = 1.1.2) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(graphite = 0.4.8) -> (memcached = 1.1.2) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(graphite = 0.4.8) -> (memcached = 1.1.2) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(jenkins = 2.0.3) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(jenkins = 2.0.3) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(jenkins = 2.0.3) -> (yum ~> 3.0)], [(logstash = 0.7.6) -> (yum >= 0.0.0)], [(logstash = 0.7.6) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(logstash = 0.7.6) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(logstash = 0.7.6) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(logstash = 0.7.6) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(logstash = 0.7.6) -> (git = 2.8.5) -> (yum < 3.0.0)], [(logstash = 0.7.6) -> (git = 2.8.5) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(logstash = 0.7.6) -> (git = 2.8.5) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(logstash = 0.7.6) -> (git = 2.8.5) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(logstash = 0.7.6) -> (git = 2.8.5) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(logstash = 0.7.6) -> (git = 2.3.0) -> (yum >= 0.0.0)], [(logstash = 0.7.6) -> (git = 2.3.0) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(logstash = 0.7.6) -> (git = 2.3.0) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(logstash = 0.7.6) -> (git = 2.3.0) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(logstash = 0.7.6) -> (git = 2.3.0) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(logstash = 0.7.6) -> (git = 4.0.0) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(logstash = 0.7.6) -> (git = 4.0.0) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(logstash = 0.7.6) -> (git = 4.0.0) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(logstash = 0.7.6) -> (git = 4.0.0) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(logstash = 0.7.6) -> (git = 4.0.0) -> (yum ~> 3.0)], [(logstash = 0.7.6) -> (git = 4.0.0) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(logstash = 0.7.6) -> (git = 2.7.0) -> (yum >= 0.0.0)], [(logstash = 0.7.6) -> (git = 2.7.0) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(logstash = 0.7.6) -> (git = 2.7.0) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(logstash = 0.7.6) -> (git = 2.7.0) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(logstash = 0.7.6) -> (git = 2.7.0) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(logstash = 0.7.6) -> (git = 2.2.0) -> (yum >= 0.0.0)], [(logstash = 0.7.6) -> (git = 2.1.2) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(logstash = 0.7.6) -> (git = 2.1.2) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(logstash = 0.7.6) -> (git = 2.1.2) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(logstash = 0.7.6) -> (git = 2.1.2) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(logstash = 0.7.6) -> (git = 2.1.2) -> (yum >= 0.0.0)], [(logstash = 0.7.6) -> (git = 2.0.0) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(logstash = 0.7.6) -> (git = 2.0.0) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(logstash = 0.7.6) -> (git = 2.0.0) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(logstash = 0.7.6) -> (git = 2.0.0) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(logstash = 0.7.6) -> (git = 2.0.0) -> (yum >= 0.0.0)], [(logstash = 0.7.6) -> (git = 1.0.0) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(logstash = 0.7.6) -> (git = 1.0.0) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(logstash = 0.7.6) -> (git = 1.0.0) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(logstash = 0.7.6) -> (git = 1.0.0) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(logstash = 0.7.6) -> (git = 1.0.0) -> (yum >= 0.0.0)], [(logstash = 0.7.6) -> (git = 0.10.0) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(logstash = 0.7.6) -> (git = 0.10.0) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(logstash = 0.7.6) -> (git = 0.10.0) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(logstash = 0.7.6) -> (git = 0.10.0) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(logstash = 0.7.6) -> (git = 0.10.0) -> (yum >= 0.0.0)], [(logstash = 0.7.6) -> (git = 0.9.0) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(logstash = 0.7.6) -> (git = 0.9.0) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(logstash = 0.7.6) -> (git = 0.9.0) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(logstash = 0.7.6) -> (git = 0.9.0) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(logstash = 0.7.6) -> (git = 0.7.0) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(logstash = 0.7.6) -> (git = 0.7.0) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(logstash = 0.7.6) -> (git = 0.7.0) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(logstash = 0.7.6) -> (git = 0.7.0) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(logstash = 0.7.6) -> (rabbitmq = 3.1.0) -> (erlang = 1.5.2) -> (yum ~> 3.0)], [(logstash = 0.7.6) -> (rabbitmq = 3.1.0) -> (erlang = 1.5.2) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(logstash = 0.7.6) -> (rabbitmq = 3.1.0) -> (erlang = 1.5.2) -> (yum-erlang_solutions = 0.2.0) -> (yum ~> 3.0)], [(logstash = 0.7.6) -> (rabbitmq = 3.1.0) -> (erlang = 1.3.6) -> (yum < 3.0.0)], [(logstash = 0.7.6) -> (rabbitmq = 3.1.0) -> (erlang = {1.3.2,1.2.0,1.1.2}) -> (yum >= 0.5.0)], [(logstash = 0.7.6) -> (rabbitmq = 1.8.0) -> (yum >= 0.5.0)], [(logstash = 0.7.6) -> (rabbitmq = 1.8.0) -> (erlang = 1.5.2) -> (yum ~> 3.0)], [(logstash = 0.7.6) -> (rabbitmq = 1.8.0) -> (erlang = 1.5.2) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(logstash = 0.7.6) -> (rabbitmq = 1.8.0) -> (erlang = 1.5.2) -> (yum-erlang_solutions = 0.2.0) -> (yum ~> 3.0)], [(logstash = 0.7.6) -> (rabbitmq = 1.8.0) -> (erlang = 1.3.6) -> (yum < 3.0.0)], [(logstash = 0.7.6) -> (rabbitmq = 1.8.0) -> (erlang = {1.3.2,1.2.0,1.1.2}) -> (yum >= 0.5.0)], [(logstash = 0.7.6) -> (rabbitmq = 1.4.0) -> (yum >= 0.5.0)], [(logstash = 0.7.6) -> (rabbitmq = 1.4.0) -> (erlang = 1.5.2) -> (yum ~> 3.0)], [(logstash = 0.7.6) -> (rabbitmq = 1.4.0) -> (erlang = 1.5.2) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(logstash = 0.7.6) -> (rabbitmq = 1.4.0) -> (erlang = 1.5.2) -> (yum-erlang_solutions = 0.2.0) -> (yum ~> 3.0)], [(logstash = 0.7.6) -> (rabbitmq = 1.4.0) -> (erlang = 1.3.6) -> (yum < 3.0.0)], [(logstash = 0.7.6) -> (rabbitmq = 1.4.0) -> (erlang = {1.3.2,1.2.0,1.1.2}) -> (yum >= 0.5.0)], [(logstash = 0.7.6) -> (python = 1.4.7) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(logstash = 0.7.6) -> (python = 1.4.4) -> (yum >= 0.0.0)], [(nginx = 2.5.1) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(nginx = 2.5.1) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(nginx = 2.5.1) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(nginx = 2.5.1) -> (runit = 1.4.0) -> (yum >= 0.0.0)], [(nginx = 2.5.1) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(php = 1.4.7) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(python = 1.4.7) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(rbenv = 1.7.1) -> (git = 2.8.5) -> (yum < 3.0.0)], [(rbenv = 1.7.1) -> (git = 2.8.5) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(rbenv = 1.7.1) -> (git = 2.8.5) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(rbenv = 1.7.1) -> (git = 2.8.5) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(rbenv = 1.7.1) -> (git = 2.8.5) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(rbenv = 1.7.1) -> (git = 2.3.0) -> (yum >= 0.0.0)], [(rbenv = 1.7.1) -> (git = 2.3.0) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(rbenv = 1.7.1) -> (git = 2.3.0) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(rbenv = 1.7.1) -> (git = 2.3.0) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(rbenv = 1.7.1) -> (git = 2.3.0) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(rbenv = 1.7.1) -> (git = 4.0.0) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(rbenv = 1.7.1) -> (git = 4.0.0) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(rbenv = 1.7.1) -> (git = 4.0.0) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(rbenv = 1.7.1) -> (git = 4.0.0) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(rbenv = 1.7.1) -> (git = 4.0.0) -> (yum ~> 3.0)], [(rbenv = 1.7.1) -> (git = 4.0.0) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(rbenv = 1.7.1) -> (git = 2.7.0) -> (yum >= 0.0.0)], [(rbenv = 1.7.1) -> (git = 2.7.0) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(rbenv = 1.7.1) -> (git = 2.7.0) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(rbenv = 1.7.1) -> (git = 2.7.0) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(rbenv = 1.7.1) -> (git = 2.7.0) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(rbenv = 1.7.1) -> (git = 2.2.0) -> (yum >= 0.0.0)], [(rbenv = 1.7.1) -> (git = 2.1.2) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(rbenv = 1.7.1) -> (git = 2.1.2) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(rbenv = 1.7.1) -> (git = 2.1.2) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(rbenv = 1.7.1) -> (git = 2.1.2) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(rbenv = 1.7.1) -> (git = 2.1.2) -> (yum >= 0.0.0)], [(rbenv = 1.7.1) -> (git = 2.0.0) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(rbenv = 1.7.1) -> (git = 2.0.0) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(rbenv = 1.7.1) -> (git = 2.0.0) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(rbenv = 1.7.1) -> (git = 2.0.0) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(rbenv = 1.7.1) -> (git = 2.0.0) -> (yum >= 0.0.0)], [(rbenv = 1.7.1) -> (git = 1.0.0) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(rbenv = 1.7.1) -> (git = 1.0.0) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(rbenv = 1.7.1) -> (git = 1.0.0) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(rbenv = 1.7.1) -> (git = 1.0.0) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(rbenv = 1.7.1) -> (git = 1.0.0) -> (yum >= 0.0.0)], [(rbenv = 1.7.1) -> (git = 0.10.0) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(rbenv = 1.7.1) -> (git = 0.10.0) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(rbenv = 1.7.1) -> (git = 0.10.0) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(rbenv = 1.7.1) -> (git = 0.10.0) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(rbenv = 1.7.1) -> (git = 0.10.0) -> (yum >= 0.0.0)], [(rbenv = 1.7.1) -> (git = 0.9.0) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(rbenv = 1.7.1) -> (git = 0.9.0) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(rbenv = 1.7.1) -> (git = 0.9.0) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(rbenv = 1.7.1) -> (git = 0.9.0) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(rbenv = 1.7.1) -> (git = 0.7.0) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(rbenv = 1.7.1) -> (git = 0.7.0) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(rbenv = 1.7.1) -> (git = 0.7.0) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(rbenv = 1.7.1) -> (git = 0.7.0) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(runit = 1.5.11) -> (yum ~> 3.0)], [(runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(statsd = 1.1.5) -> (git = 2.8.5) -> (yum < 3.0.0)], [(statsd = 1.1.5) -> (git = 2.8.5) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(statsd = 1.1.5) -> (git = 2.8.5) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(statsd = 1.1.5) -> (git = 2.8.5) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(statsd = 1.1.5) -> (git = 2.8.5) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(statsd = 1.1.5) -> (git = 2.3.0) -> (yum >= 0.0.0)], [(statsd = 1.1.5) -> (git = 2.3.0) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(statsd = 1.1.5) -> (git = 2.3.0) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(statsd = 1.1.5) -> (git = 2.3.0) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(statsd = 1.1.5) -> (git = 2.3.0) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(statsd = 1.1.5) -> (git = 4.0.0) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(statsd = 1.1.5) -> (git = 4.0.0) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(statsd = 1.1.5) -> (git = 4.0.0) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(statsd = 1.1.5) -> (git = 4.0.0) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(statsd = 1.1.5) -> (git = 4.0.0) -> (yum ~> 3.0)], [(statsd = 1.1.5) -> (git = 4.0.0) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(statsd = 1.1.5) -> (git = 2.7.0) -> (yum >= 0.0.0)], [(statsd = 1.1.5) -> (git = 2.7.0) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(statsd = 1.1.5) -> (git = 2.7.0) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(statsd = 1.1.5) -> (git = 2.7.0) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(statsd = 1.1.5) -> (git = 2.7.0) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(statsd = 1.1.5) -> (git = 2.2.0) -> (yum >= 0.0.0)], [(statsd = 1.1.5) -> (git = 2.1.2) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(statsd = 1.1.5) -> (git = 2.1.2) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(statsd = 1.1.5) -> (git = 2.1.2) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(statsd = 1.1.5) -> (git = 2.1.2) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(statsd = 1.1.5) -> (git = 2.1.2) -> (yum >= 0.0.0)], [(statsd = 1.1.5) -> (git = 2.0.0) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(statsd = 1.1.5) -> (git = 2.0.0) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(statsd = 1.1.5) -> (git = 2.0.0) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(statsd = 1.1.5) -> (git = 2.0.0) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(statsd = 1.1.5) -> (git = 2.0.0) -> (yum >= 0.0.0)], [(statsd = 1.1.5) -> (git = 1.0.0) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(statsd = 1.1.5) -> (git = 1.0.0) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(statsd = 1.1.5) -> (git = 1.0.0) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(statsd = 1.1.5) -> (git = 1.0.0) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(statsd = 1.1.5) -> (git = 1.0.0) -> (yum >= 0.0.0)], [(statsd = 1.1.5) -> (git = 0.10.0) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(statsd = 1.1.5) -> (git = 0.10.0) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(statsd = 1.1.5) -> (git = 0.10.0) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(statsd = 1.1.5) -> (git = 0.10.0) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(statsd = 1.1.5) -> (git = 0.10.0) -> (yum >= 0.0.0)], [(statsd = 1.1.5) -> (git = 0.9.0) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(statsd = 1.1.5) -> (git = 0.9.0) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(statsd = 1.1.5) -> (git = 0.9.0) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(statsd = 1.1.5) -> (git = 0.9.0) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(statsd = 1.1.5) -> (git = 0.7.0) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(statsd = 1.1.5) -> (git = 0.7.0) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(statsd = 1.1.5) -> (git = 0.7.0) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(statsd = 1.1.5) -> (git = 0.7.0) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(statsd = 1.1.5) -> (nodejs = 1.3.0) -> (yum >= 0.0.0)], [(statsd = 1.1.5) -> (runit = 1.5.11) -> (yum ~> 3.0)], [(statsd = 1.1.5) -> (runit = 1.5.11) -> (yum-epel = 0.3.4) -> (yum ~> 3.0)], [(statsd = 1.1.5) -> (runit = 1.1.4) -> (yum >= 0.0.0)], [(statsd = 1.1.5) -> (runit = 1.4.6) -> (yum < 3.0.0)], [(virtualbox-cookbook = 1.0.3) -> (yum ~> 3.1)], [(yum = 2.4.5)]
Demand that cannot be met: (jenkins = 2.0.3)
Artifacts for which there are conflicting dependencies: yum = 2.4.5 -> []
```

reset commented:
@danielsdeleo overall the code looks great - some comments from your dev pass that you could go and strip out and that's about it!

Yeah, I expected the tests to match more closely the old Ruby solver and not Gecode, sigh. I think it's appropriate to include this as long as we are clear in the Berkshelf docs that if a solution does not match on the server based on what your client resolver thought, you should switch to gecode for compat reasons.

I like the output of a failed resolution MUCH more than what we were getting from gecode.

Aside from getting the remaining tests passing, is there anything left that you'd like to do before merging this in? I'm down to update docs and release a new major version of Solve.

danielsdeleo commented:
@reset Things remaining to do that I can remember:

* I think the code that switches engines could use a bit of improvement so that it's easier to write the code in berks that does the actual engine picking. In particular, it shouldn't just be a generic LoadError if you try to use the gecode solver and it's not available.
* Remove runtime dep on dep_selector. I'm thinking to specify it as a development dependency (?)
* Rename Solver to GecodeSolver
* make sure the serializer works with the ruby solver

reset commented:
@danielsdeleo yeah I'm also good with making the Gecode solver a sort of "plugin" or "extension" to Solve which would allow you to split it into a separate gem and instead ask people to install that if they'd like the option of switching solver backends.

danielsdeleo commented:
@reset quick question: The serializer right now only supports one kind of solver. So I can either add a `type` field to the json, or I could make a `Solver::Problem` class that has the required fields. Do you have a preference for one over the other?

reset commented:
@danielsdeleo I think that either approach is just fine, I would decide which based on the cost of work

danielsdeleo commented:
Okay, I think this is good to go. I added an `::activate` class method to the solvers, which will raise `EngineNotAvailable` if there's a missing dep. So berks should be able to do `Solve.engine = :gecode` and then rescue and print a warning for the case that someone gem installed berks and hasn't set a preference for the ruby solver.

I mucked with the gemspec and Gemfile a bit, making test dependencies rubygems-level development dependencies. I did that so I could add a `NoGecode.gemfile` and corresponding travis job. That ensures that if we accidentally create a dependency on the dep selector stuff where we don't mean to, a travis job will fail.

Also, I opted go for the `Solve::Problem` class in the serializer. There's a convenience method to create one directly from a solver object so it's only one more line of code to turn a solver into JSON.

reset commented on lib/solve/graph.rb:
```diff
@@ -2,6 +2,7 @@ module Solve
   class Graph
     def initialize
       @artifacts = {}
+      #@artifacts_by_name = Hash.new { |hash, key| hash[key] = [] }
```
Don't think we need this comment

danielsdeleo replied:
Oops, good catch, I'll fix this up.

reset commented:
@danielsdeleo this looks great! If there's nothing else that you'd like to add we can merge it in and make a new major release.

danielsdeleo commented:
Yep, I don't have anything else I want to add. Let's go!
